crates/block-producer/src/mempool/state_dag.rs

```diff
@@ -0,0 +1,442 @@
+#![allow(
+    dead_code,
+    clippy::unused_self,
+    unused_variables,
+    clippy::needless_pass_by_value,
+    reason = "wip"
+)]
+
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+
+use miden_objects::batch::{BatchId, ProvenBatch};
+use miden_objects::block::BlockNumber;
+use miden_objects::note::Nullifier;
+use miden_objects::transaction::TransactionId;
+
+use crate::domain::transaction::AuthenticatedTransaction;
+use crate::errors::AddTransactionError;
+use crate::mempool::{BatchBudget, BlockBudget, BudgetStatus};
+
+// STATE DAG
+// ================================================================================================
+
+/// Tracks the inflight state of the mempool and models the relationship of the transactions,
+/// batches and blocks as a DAG.
+///
+/// This model allows it to propose batches and blocks from the available transactions and proven
+/// batches respectively. New transactions are validated against the current inflight state.
+#[derive(Clone, Debug, PartialEq)]
+pub struct StateDag {
+    /// The DAG nodes describing the transactions, batches and blocks currently in the mempool.
+    ///
+    /// Only active nodes are stored e.g. a proposed batch will _replace_ all its transaction
+    /// nodes. Relationships between the nodes are inferred using the state which essentially
+    /// describes the edges of the DAG.
+    ///
+    /// This is kept in-sync with the state property i.e. the state is an aggregation of the
+    /// nodes' state.
+    nodes: Nodes,
+
+    /// Describes the committed, created and consumed state of the nodes in the graph. The creating
+    /// and consuming [`NodeId's`] are stored alongside the state, which allows this to act as the
+    /// edges of the DAG.
+    state: InflightState,
+
+    /// Number of committed block's whose state we retain.
+    ///
+    /// This provides an overlap with the committed state from the store, giving new data's
+    /// authentication inputs a grace period before it is submitted to the mempool. Without this,
+    /// fetching the inputs from the store would race against new block being committed.
+    state_retention: NonZeroUsize,
```

Comment:
Could we make this deterministic through means of a feedback channel from the block import / application pipeline?